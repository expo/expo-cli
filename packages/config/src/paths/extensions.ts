import invariant from 'invariant';

export type LanguageOptions = {
  isTS: boolean;
  isModern: boolean;
  isReact: boolean;
};

export function getExtensions(
  platforms: string[],
  extensions: string[],
  workflows: string[]
): string[] {
  // In the past we used spread operators to collect the values so now we enforce type safety on them.
  invariant(Array.isArray(platforms), 'Expected: `platforms: string[]`');
  invariant(Array.isArray(extensions), 'Expected: `extensions: string[]`');
  invariant(Array.isArray(workflows), 'Expected: `workflows: string[]`');

  const fileExtensions = [];
  // support .expo files
  for (const workflow of [...workflows, '']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      // Support both TypeScript and JavaScript
      for (const extension of extensions) {
        fileExtensions.push([platform, workflow, extension].filter(Boolean).join('.'));
      }
    }
  }
  return fileExtensions;
}

export function getLanguageExtensionsInOrder({
  isTS,
  isModern,
  isReact,
}: LanguageOptions): string[] {
  // @ts-ignore: filter removes false type
  const addLanguage = (lang: string): string[] => [lang, isReact && `${lang}x`].filter(Boolean);

  // Support JavaScript
  let extensions = addLanguage('js');

  if (isModern) {
    extensions.unshift('mjs');
  }
  if (isTS) {
    extensions = [...addLanguage('ts'), ...extensions];
  }

  return extensions;
}

export function getManagedExtensions(
  platforms: string[],
  languageOptions: LanguageOptions = { isTS: true, isModern: true, isReact: true }
): string[] {
  const fileExtensions = getExtensions(platforms, getLanguageExtensionsInOrder(languageOptions), [
    'expo',
  ]);
  // Always add these last
  _addMiscellaneousExtensions(fileExtensions);
  return fileExtensions;
}

export function getBareExtensions(
  platforms: string[],
  languageOptions: LanguageOptions = { isTS: true, isModern: true, isReact: true }
): string[] {
  const fileExtensions = getExtensions(
    platforms,
    getLanguageExtensionsInOrder(languageOptions),
    []
  );
  // Always add these last
  _addMiscellaneousExtensions(fileExtensions);
  return fileExtensions;
}

function _addMiscellaneousExtensions(fileExtensions: string[]): string[] {
  // Always add these with no platform extension
  // In the future we may want to add platform and workspace extensions to json.
  fileExtensions.push('json');
  fileExtensions.push('wasm');
  return fileExtensions;
}
