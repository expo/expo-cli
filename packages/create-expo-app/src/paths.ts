import { homedir } from 'os';
import * as path from 'path';

// copied from https://github.com/expo/eas-cli/blob/f0c958e58bc7aa90ee8f822e075d40703563708e/packages/eas-cli/src/utils/paths.ts

// The ~/.expo directory is used to store authentication sessions,
// which are shared between EAS CLI and Expo CLI.
function dotExpoHomeDirectory(): string {
  const home = homedir();
  if (!home) {
    throw new Error(
      "Can't determine your home directory; make sure your $HOME environment variable is set."
    );
  }

  let dirPath;
  if (process.env.EXPO_STAGING) {
    dirPath = path.join(home, '.expo-staging');
  } else if (process.env.EXPO_LOCAL) {
    dirPath = path.join(home, '.expo-local');
  } else {
    dirPath = path.join(home, '.expo');
  }
  return dirPath;
}

const getStateJsonPath = (): string => path.join(dotExpoHomeDirectory(), 'state.json');

export { getStateJsonPath };
