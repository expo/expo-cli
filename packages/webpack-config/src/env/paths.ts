/* eslint-env node */
import { ExpoConfig, getConfig, getWebOutputPath } from '@expo/config';
import {
  ensureSlash,
  getAbsolutePathWithProjectRoot,
  getEntryPoint,
  getModulesPath,
  getPossibleProjectRoot,
} from '@expo/config/paths';
import fs from 'fs';
import path from 'path';
import url from 'url';

import { Environment, FilePaths, Mode } from '../types';
import getMode from './getMode';
import { getConfigMode } from './getConfigMode';

function parsePaths(projectRoot: string, mode: Mode, nativeAppManifest?: ExpoConfig): FilePaths {
  const inputProjectRoot = projectRoot || getPossibleProjectRoot();

  function absolute(...pathComponents: string[]): string {
    return getAbsolutePathWithProjectRoot(inputProjectRoot, ...pathComponents);
  }

  const packageJsonPath = absolute('package.json');
  const modulesPath = getModulesPath(inputProjectRoot);
  const productionPath = absolute(getWebOutputPath(nativeAppManifest));

  function templatePath(filename: string = ''): string {
    const overridePath = absolute('web', filename);
    if (fs.existsSync(overridePath)) {
      return overridePath;
    }
    return path.join(__dirname, '../../web-default', filename);
  }

  function getProductionPath(...props: string[]): string {
    return path.resolve(productionPath, ...props);
  }

  function getIncludeModule(...props: string[]): string {
    return path.resolve(modulesPath, ...props);
  }

  let appMain: string | null = null;
  try {
    appMain = getEntryPoint(inputProjectRoot, ['./index', './src/index'], ['web']);
  } catch (_) {
    // ignore the error
  }

  return {
    absolute,
    includeModule: getIncludeModule,
    packageJson: packageJsonPath,
    root: path.resolve(inputProjectRoot),
    appMain,
    modules: modulesPath,
    servedPath: getServedPath(inputProjectRoot, mode),
    template: {
      get: templatePath,
      folder: templatePath(),
      indexHtml: templatePath('index.html'),
      manifest: templatePath('manifest.json'),
      serveJson: templatePath('serve.json'),
      favicon: templatePath('favicon.ico'),
      serviceWorker: templatePath('expo-service-worker.js'),
      registerServiceWorker: templatePath('register-service-worker.js'),
    },
    production: {
      get: getProductionPath,
      folder: getProductionPath(),
      indexHtml: getProductionPath('index.html'),
      manifest: getProductionPath('manifest.json'),
      serveJson: getProductionPath('serve.json'),
      favicon: getProductionPath('favicon.ico'),
      serviceWorker: getProductionPath('expo-service-worker.js'),
      registerServiceWorker: getProductionPath('register-service-worker.js'),
    },
  };
}

/**
 * Sync method for getting default paths used throughout the Webpack config.
 * This is useful for Next.js which doesn't support async Webpack configs.
 *
 * @param projectRoot
 * @category env
 */
export function getPaths(projectRoot: string, mode?: string): FilePaths {
  const configMode = getConfigMode(getMode({ mode }));
  const { exp } = getConfig(projectRoot, {
    skipSDKVersionRequirement: true,
    mode: configMode,
  });
  return parsePaths(projectRoot, configMode, exp);
}

/**
 * Async method for getting default paths used throughout the Webpack config.
 *
 * @param projectRoot
 * @category env
 */
export async function getPathsAsync(projectRoot: string, mode?: string): Promise<FilePaths> {
  const configMode = getConfigMode(getMode({ mode }));

  let exp;
  try {
    exp = getConfig(projectRoot, { skipSDKVersionRequirement: true, mode: configMode }).exp;
  } catch (error) {}
  return parsePaths(projectRoot, configMode, exp);
}

/**
 * Get paths dictating where the app is served regardless of the current Webpack mode.
 *
 * @param projectRoot
 * @category env
 */
export function getServedPath(projectRoot: string, mode?: string): string {
  const configMode = getConfigMode(getMode({ mode }));

  const { pkg } = getConfig(projectRoot, {
    skipSDKVersionRequirement: true,
    mode: configMode,
  });
  const envPublicUrl = process.env.WEB_PUBLIC_URL;

  // We use `WEB_PUBLIC_URL` environment variable or "homepage" field to infer
  // "public path" at which the app is served.
  // Webpack needs to know it to put the right <script> hrefs into HTML even in
  // single-page apps that may serve index.html for nested URLs like /todos/42.
  // We can't use a relative path in HTML because we don't want to load something
  // like /todos/42/static/js/bundle.7289d.js. We have to know the root.
  const publicUrl = envPublicUrl || pkg.homepage;
  const servedUrl = envPublicUrl || (publicUrl ? url.parse(publicUrl).pathname : '/');
  return ensureSlash(servedUrl!, true);
}

/**
 * Get paths dictating where the app is served. In development mode this returns default values.
 *
 * @param env
 * @category env
 */
export function getPublicPaths(
  env: Pick<Environment, 'mode' | 'projectRoot'>
): {
  /**
   * Webpack uses `publicPath` to determine where the app is being served from.
   * It requires a trailing slash, or the file assets will get an incorrect path.
   * In development, we always serve from the root. This makes config easier.
   */
  publicPath: string;

  /**
   * `publicUrl` is just like `publicPath`, but we will provide it to our app
   * as %WEB_PUBLIC_URL% in `index.html` and `process.env.WEB_PUBLIC_URL` in JavaScript.
   * Omit trailing slash as %WEB_PUBLIC_URL%/xyz looks better than %WEB_PUBLIC_URL%xyz.
   */
  publicUrl: string;
} {
  const parsedMode = getMode(env);
  if (parsedMode === 'production') {
    const publicPath = getServedPath(env.projectRoot, parsedMode);
    return {
      publicPath,
      publicUrl: publicPath.slice(0, -1),
    };
  }

  return { publicUrl: '', publicPath: '/' };
}

/**
 * Get the output folder path. Defaults to `web-build`.
 *
 * @param projectRoot
 * @category env
 */
export function getProductionPath(projectRoot: string, mode?: string): string {
  const configMode = getConfigMode(getMode({ mode }));

  const { exp } = getConfig(projectRoot, {
    skipSDKVersionRequirement: true,
    mode: configMode,
  });
  return getAbsolutePathWithProjectRoot(projectRoot, getWebOutputPath(exp));
}

/**
 * Get an absolute path relative to the project root while accounting for remote paths (`https://`).
 *
 * @param projectRoot
 * @category env
 */
export function getAbsolute(projectRoot: string, ...pathComponents: string[]): string {
  const inputProjectRoot = projectRoot || getPossibleProjectRoot();
  return getAbsolutePathWithProjectRoot(inputProjectRoot, ...pathComponents);
}
