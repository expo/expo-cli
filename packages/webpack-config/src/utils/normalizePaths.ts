/**
 * Converts absolute paths to relative paths for testing purposes.
 *
 * @param initial
 * @param transformString
 * @internal
 */
export default function normalizePaths(
  initial: any,
  transformString: (value: string) => string
): any {
  if (initial == null) {
    return initial;
  } else if (typeof initial === 'string') {
    return transformString(initial);
  } else if (Array.isArray(initial)) {
    return initial.map(value => normalizePaths(value, transformString));
  } else if (typeof initial === 'object') {
    let result: { [key: string]: any } = {};
    for (const prop of Object.keys(initial)) {
      result[prop] = normalizePaths(initial[prop], transformString);
    }
    return result;
  } else {
    return initial;
  }
}
