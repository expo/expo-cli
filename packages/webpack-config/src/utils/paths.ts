import { ExpoConfig, getWebOutputPath, readConfigJson, readConfigJsonAsync } from '@expo/config';
import findWorkspaceRoot from 'find-yarn-workspace-root';
import fs from 'fs';
import path from 'path';
import url from 'url';

import { Environment, FilePaths } from '../types';
import getMode from './getMode';

const possibleMainFiles = [
  'index.web.ts',
  'index.ts',
  'index.web.tsx',
  'index.tsx',
  'src/index.web.ts',
  'src/index.ts',
  'src/index.web.tsx',
  'src/index.tsx',
  'index.web.js',
  'index.js',
  'index.web.jsx',
  'index.jsx',
  'src/index.web.js',
  'src/index.js',
  'src/index.web.jsx',
  'src/index.jsx',
  'node_modules/expo/AppEntry.js',
];

function ensureSlash(inputPath: string, needsSlash: boolean): string {
  const hasSlash = inputPath.endsWith('/');
  if (hasSlash && !needsSlash) {
    return inputPath.substr(0, inputPath.length - 1);
  } else if (!hasSlash && needsSlash) {
    return `${inputPath}/`;
  } else {
    return inputPath;
  }
}

export function getAbsolutePathWithProjectRoot(
  projectRoot: string,
  ...pathComponents: string[]
): string {
  // Simple check if we are dealing with an URL
  if (pathComponents && pathComponents.length === 1 && pathComponents[0].startsWith('http')) {
    return pathComponents[0];
  }

  return path.resolve(projectRoot, ...pathComponents);
}

export function getPossibleProjectRoot(): string {
  return fs.realpathSync(process.cwd());
}

export function getModulesPath(projectRoot: string): string {
  const workspaceRoot = findWorkspaceRoot(path.resolve(projectRoot)); // Absolute path or null
  if (workspaceRoot) {
    return path.resolve(workspaceRoot, 'node_modules');
  }

  return path.resolve(projectRoot, 'node_modules');
}

function parsePaths(projectRoot: string, nativeAppManifest?: ExpoConfig): FilePaths {
  const inputProjectRoot = projectRoot || getPossibleProjectRoot();

  function absolute(...pathComponents: string[]): string {
    return getAbsolutePathWithProjectRoot(inputProjectRoot, ...pathComponents);
  }

  const packageJsonPath = absolute('package.json');
  const modulesPath = getModulesPath(inputProjectRoot);
  const productionPath = absolute(getWebOutputPath(nativeAppManifest));

  function templatePath(filename: string = ''): string {
    const overridePath = absolute('web', filename);
    if (fs.existsSync(overridePath)) {
      return overridePath;
    }
    return path.join(__dirname, '../../web-default', filename);
  }

  function getProductionPath(...props: string[]): string {
    return path.resolve(productionPath, ...props);
  }

  function getIncludeModule(...props: string[]): string {
    return path.resolve(modulesPath, ...props);
  }

  return {
    absolute,
    includeModule: getIncludeModule,
    packageJson: packageJsonPath,
    root: path.resolve(inputProjectRoot),
    appMain: getEntryPoint(inputProjectRoot),
    modules: modulesPath,
    servedPath: getServedPath(inputProjectRoot),
    template: {
      get: templatePath,
      folder: templatePath(),
      indexHtml: templatePath('index.html'),
      manifest: templatePath('manifest.json'),
      serveJson: templatePath('serve.json'),
      favicon: templatePath('favicon.ico'),
      serviceWorker: templatePath('expo-service-worker.js'),
      registerServiceWorker: templatePath('register-service-worker.js'),
    },
    production: {
      get: getProductionPath,
      folder: getProductionPath(),
      indexHtml: getProductionPath('index.html'),
      manifest: getProductionPath('manifest.json'),
      serveJson: getProductionPath('serve.json'),
      favicon: getProductionPath('favicon.ico'),
      serviceWorker: getProductionPath('expo-service-worker.js'),
      registerServiceWorker: getProductionPath('register-service-worker.js'),
    },
  };
}

export function getEntryPoint(projectRoot: string): string | null {
  const { exp, pkg } = readConfigJson(projectRoot, true, true);

  /**
   *  The main file is resolved like so:
   * * `app.json` -> `expo.entryPoint`
   * * `package.json` -> `"main"`
   * * `possibleMainFiles`
   */
  if (exp && exp.entryPoint && typeof exp.entryPoint === 'string') {
    return getAbsolutePathWithProjectRoot(projectRoot, exp.entryPoint);
  } else if (pkg) {
    const { main } = pkg;
    if (main && typeof main === 'string') {
      return getAbsolutePathWithProjectRoot(projectRoot, main);
    }
    // Adds support for create-react-app (src/index.js) and react-native-cli (index.js) which don't define a main.
    for (const fileName of possibleMainFiles) {
      const filePath = getAbsolutePathWithProjectRoot(projectRoot, fileName);
      if (fs.existsSync(filePath)) {
        return filePath;
      }
    }
  }
  return null;
}

export function getPaths(projectRoot: string): FilePaths {
  const { exp } = readConfigJson(projectRoot, true, true);
  return parsePaths(projectRoot, exp);
}

export async function getPathsAsync(projectRoot: string): Promise<FilePaths> {
  let exp;
  try {
    exp = (await readConfigJsonAsync(projectRoot, true, true)).exp;
  } catch (error) {}
  return parsePaths(projectRoot, exp);
}

export function getServedPath(projectRoot: string): string {
  const { pkg } = readConfigJson(projectRoot, true, true);
  const envPublicUrl = process.env.WEB_PUBLIC_URL;

  // We use `WEB_PUBLIC_URL` environment variable or "homepage" field to infer
  // "public path" at which the app is served.
  // Webpack needs to know it to put the right <script> hrefs into HTML even in
  // single-page apps that may serve index.html for nested URLs like /todos/42.
  // We can't use a relative path in HTML because we don't want to load something
  // like /todos/42/static/js/bundle.7289d.js. We have to know the root.
  const publicUrl = envPublicUrl || pkg.homepage;
  const servedUrl = envPublicUrl || (publicUrl ? url.parse(publicUrl).pathname : '/');
  return ensureSlash(servedUrl!, true);
}

export function getPublicPaths({
  projectRoot,
  ...env
}: Environment): {
  /**
   * Webpack uses `publicPath` to determine where the app is being served from.
   * It requires a trailing slash, or the file assets will get an incorrect path.
   * In development, we always serve from the root. This makes config easier.
   */
  publicPath: string;

  /**
   * `publicUrl` is just like `publicPath`, but we will provide it to our app
   * as %WEB_PUBLIC_URL% in `index.html` and `process.env.WEB_PUBLIC_URL` in JavaScript.
   * Omit trailing slash as %WEB_PUBLIC_URL%/xyz looks better than %WEB_PUBLIC_URL%xyz.
   */
  publicUrl: string;
} {
  if (getMode(env) === 'production') {
    const publicPath = getServedPath(projectRoot);
    return {
      publicPath,
      publicUrl: publicPath.slice(0, -1),
    };
  }

  return { publicUrl: '', publicPath: '/' };
}

export function getProductionPath(projectRoot: string): string {
  const { exp } = readConfigJson(projectRoot, true, true);
  return getAbsolutePathWithProjectRoot(projectRoot, getWebOutputPath(exp));
}
