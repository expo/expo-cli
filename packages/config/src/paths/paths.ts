import fs from 'fs-extra';
import path from 'path';
import findWorkspaceRoot from 'find-yarn-workspace-root';

import resolveFrom from 'resolve-from';
import { getConfig } from '../Config';
import { resolveModule } from '../Modules';
import { getManagedExtensions } from './extensions';

// https://github.com/facebook/create-react-app/blob/9750738cce89a967cc71f28390daf5d4311b193c/packages/react-scripts/config/paths.js#L22
export function ensureSlash(inputPath: string, needsSlash: boolean): string {
  const hasSlash = inputPath.endsWith('/');
  if (hasSlash && !needsSlash) {
    return inputPath.substr(0, inputPath.length - 1);
  } else if (!hasSlash && needsSlash) {
    return `${inputPath}/`;
  } else {
    return inputPath;
  }
}

export function getModulesPath(projectRoot: string): string {
  const workspaceRoot = findWorkspaceRoot(path.resolve(projectRoot)); // Absolute path or null
  if (workspaceRoot) {
    return path.resolve(workspaceRoot, 'node_modules');
  }

  return path.resolve(projectRoot, 'node_modules');
}

export function getPossibleProjectRoot(): string {
  return fs.realpathSync(process.cwd());
}

export function getAbsolutePathWithProjectRoot(
  projectRoot: string,
  ...pathComponents: string[]
): string {
  // Simple check if we are dealing with an URL
  if (pathComponents && pathComponents.length === 1 && pathComponents[0].startsWith('http')) {
    return pathComponents[0];
  }

  return path.resolve(projectRoot, ...pathComponents);
}

export function getEntryPoint(
  projectRoot: string,
  entryFiles: string[],
  platforms: string[]
): string | null {
  const extensions = getManagedExtensions(platforms);
  return getEntryPointWithExtensions(projectRoot, entryFiles, extensions);
}

// Used to resolve the main entry file for a project.
export function getEntryPointWithExtensions(
  projectRoot: string,
  entryFiles: string[],
  extensions: string[]
): string {
  const { exp, pkg } = getConfig(projectRoot, { skipSDKVersionRequirement: true });

  // This will first look in the `app.json`s `expo.entryPoint` field for a potential main file.
  // We check the Expo config first in case you want your project to start differently with Expo then in a standalone environment.
  if (exp && exp.entryPoint && typeof exp.entryPoint === 'string') {
    // If the field exists then we want to test it against every one of the supplied extensions
    // to ensure the bundler resolves the same way.
    let entry = getFileWithExtensions(projectRoot, exp.entryPoint, extensions);
    if (!entry) {
      // Allow for paths like: `{ "main": "expo/AppEntry" }`
      entry = resolveFromSilentWithExtensions(projectRoot, exp.entryPoint, extensions);
      if (!entry)
        throw new Error(
          `Cannot resolve entry file: The \`expo.entryPoint\` field defined in your \`app.json\` points to a non-existent path.`
        );
    }
    return entry;
  } else if (pkg) {
    // If the config doesn't define a custom entry then we want to look at the `package.json`s `main` field, and try again.
    const { main } = pkg;
    if (main && typeof main === 'string') {
      // Testing the main field against all of the provided extensions - for legacy reasons we can't use node module resolution as the package.json allows you to pass in a file without a relative path and expect it as a relative path.
      let entry = getFileWithExtensions(projectRoot, main, extensions);
      if (!entry) {
        // Allow for paths like: `{ "main": "expo/AppEntry" }`
        entry = resolveFromSilentWithExtensions(projectRoot, main, extensions);
        if (!entry)
          throw new Error(
            `Cannot resolve entry file: The \`main\` field defined in your \`package.json\` points to a non-existent path.`
          );
      }
      return entry;
    }
  }

  // Now we will start looking for a default entry point using the provided `entryFiles` argument.
  // This will add support for create-react-app (src/index.js) and react-native-cli (index.js) which don't define a main.
  for (const fileName of entryFiles) {
    const entry = resolveFromSilentWithExtensions(projectRoot, fileName, extensions);
    if (entry) return entry;
  }

  try {
    // If none of the default files exist then we will attempt to use the main Expo entry point.
    // This requires `expo` to be installed in the project to work as it will use `node_module/expo/AppEntry.js`
    // Doing this enables us to create a bare minimum Expo project.

    // TODO(Bacon): We may want to do a check against `./App` and `expo` in the `package.json` `dependencies` as we can more accurately ensure that the project is expo-min without needing the modules installed.
    return resolveModule('expo/AppEntry', projectRoot, exp);
  } catch (_) {
    throw new Error(
      `The project entry file could not be resolved. Please either define it in the \`package.json\` (main), \`app.json\` (expo.entryPoint), create an \`index.js\`, or install the \`expo\` package.`
    );
  }
}

// Resolve from but with the ability to resolve like a bundler
export function resolveFromSilentWithExtensions(
  fromDirectory: string,
  moduleId: string,
  extensions: string[]
): string | null {
  for (const extension of extensions) {
    const modulePath = resolveFrom.silent(fromDirectory, `${moduleId}.${extension}`);
    if (modulePath && modulePath.endsWith(extension)) {
      return modulePath;
    }
  }
  return resolveFrom.silent(fromDirectory, moduleId) || null;
}

// Statically attempt to resolve a module but with the ability to resolve like a bundler.
// This won't use node module resolution.
export function getFileWithExtensions(
  fromDirectory: string,
  moduleId: string,
  extensions: string[]
): string | null {
  const modulePath = path.join(fromDirectory, moduleId);
  if (fs.existsSync(modulePath)) {
    return modulePath;
  }
  for (const extension of extensions) {
    const modulePath = path.join(fromDirectory, `${moduleId}.${extension}`);
    if (fs.existsSync(modulePath)) {
      return modulePath;
    }
  }
  return null;
}
