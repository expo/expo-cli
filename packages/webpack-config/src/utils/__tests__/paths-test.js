import path from 'path';
import { getPathsAsync, getPaths } from '../paths';
import normalizePaths from '../normalizePaths';

const projectRoot = path.resolve(__dirname, '../../../tests/basic');
const projectRootCustomHomepage = path.resolve(__dirname, '../../../tests/custom-homepage');

function defaultNormalize(locations) {
  return normalizePaths(locations, value => value.split('packages/webpack-config/').pop());
}

beforeEach(() => {
  delete process.env.WEB_PUBLIC_URL;
});

it(`has consistent defaults`, async () => {
  const locations = await getPathsAsync(projectRoot);

  const normalized = defaultNormalize(locations);
  expect(normalized).toMatchSnapshot();

  expect(locations.servedPath).toBe('/');
});

it(`matches sync and async results`, async () => {
  const locations = getPaths(projectRoot);
  const normalized = defaultNormalize(locations);

  const locationsAsync = await getPathsAsync(projectRoot);
  const normalizedAsync = defaultNormalize(locationsAsync);

  expect(JSON.stringify(normalized)).toBe(JSON.stringify(normalizedAsync));
});

it(`uses a custom public path from WEB_PUBLIC_URL over the homepage field from package.json`, async () => {
  process.env.WEB_PUBLIC_URL = 'WEB_PUBLIC_URL-defined';
  const locations = await getPathsAsync(projectRootCustomHomepage);
  expect(locations.servedPath).toBe('WEB_PUBLIC_URL-defined/');
});

it(`uses a custom public path from the homepage field of a project's package.json`, async () => {
  const locations = await getPathsAsync(projectRootCustomHomepage);
  expect(locations.servedPath).toMatchSnapshot();
});
// TODO: Bacon: Add test for resolving entry point
// TODO: Bacon: Add test for custom config paths
