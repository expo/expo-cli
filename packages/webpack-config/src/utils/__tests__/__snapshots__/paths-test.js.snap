// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has consistent defaults 1`] = `
Object {
  "absolute": [Function],
  "appMain": "tests/basic/index",
  "includeModule": [Function],
  "modules": "tests/basic/node_modules",
  "packageJson": "tests/basic/package.json",
  "production": Object {
    "favicon": "tests/basic/web-build/favicon.ico",
    "folder": "tests/basic/web-build",
    "get": [Function],
    "indexHtml": "tests/basic/web-build/index.html",
    "manifest": "tests/basic/web-build/manifest.json",
    "serveJson": "tests/basic/web-build/serve.json",
    "serviceWorker": "tests/basic/web-build/expo-service-worker.js",
  },
  "root": "tests/basic",
  "servedPath": "/",
  "template": Object {
    "favicon": "web-default/favicon.ico",
    "folder": "web-default",
    "get": [Function],
    "indexHtml": "web-default/index.html",
    "manifest": "web-default/manifest.json",
    "serveJson": "web-default/serve.json",
    "serviceWorker": "web-default/expo-service-worker.js",
  },
}
`;

exports[`uses a custom public path from the homepage field of a project's package.json 1`] = `"/homepage-defined/"`;

exports[`works with no app.json 1`] = `
Object {
  "absolute": [Function],
  "appMain": "tests/minimum/index.js",
  "includeModule": [Function],
  "modules": "tests/minimum/node_modules",
  "packageJson": "tests/minimum/package.json",
  "production": Object {
    "favicon": "tests/minimum/web-build/favicon.ico",
    "folder": "tests/minimum/web-build",
    "get": [Function],
    "indexHtml": "tests/minimum/web-build/index.html",
    "manifest": "tests/minimum/web-build/manifest.json",
    "serveJson": "tests/minimum/web-build/serve.json",
    "serviceWorker": "tests/minimum/web-build/expo-service-worker.js",
  },
  "root": "tests/minimum",
  "servedPath": "/",
  "template": Object {
    "favicon": "web-default/favicon.ico",
    "folder": "web-default",
    "get": [Function],
    "indexHtml": "web-default/index.html",
    "manifest": "web-default/manifest.json",
    "serveJson": "web-default/serve.json",
    "serviceWorker": "web-default/expo-service-worker.js",
  },
}
`;
